### PR TITLE
Make sure the Itemid get calls are always casted to integer

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Make sure the Itemid filter always casts to integer, when the itemid get passed an array of items the current [getint filter (`int`, `integer`)](https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Filter/InputFilter.php#L117) returns an array but the itemid should never be an array.

### Testing Instructions

Make sure the usual Frontend and backend still works and the correct itemid is in the html classes.

### Expected result

We type cast the itemid to integer

### Actual result

Passing an array would result into the itemId not to be an actual ID any more but an array.

### Documentation Changes Required

None cc @SniperSister @HLeithner 

Reported by: https://twitter.com/paulosyibelo

### CODE REVIEW

File: administrator/templates/isis/error.php [hunk 1]
```diff
@@ -27,11 +27,11 @@
 $mainPageUri = $frontEndUri->toString();
 
 // Detecting Active Variables
-$option   = $input->get('option', '');
-$view     = $input->get('view', '');
-$layout   = $input->get('layout', '');
-$task     = $input->get('task', '');
-$itemid   = $input->get('Itemid', 0, 'int');
+$option   = $input->getCmd('option', '');
```
Comment (SharkyKZ) [inline administrator/templates/isis/error.php]: This is unnecessary. `getCmd()` is just a slow version of `get()`.

Comment (zero-24) [reply administrator/templates/isis/error.php]: Fixed: https://github.com/joomla/joomla-cms/pull/28481/commits/5957737c081937de90ff09a4b8a0fa492ad9f021

File: libraries/src/Helper/ModuleHelper.php [hunk 2]
```diff
@@ -451,7 +451,7 @@ public static function getModuleList()
 	public static function cleanModuleList($modules)
 	{
 		// Apply negative selections and eliminate duplicates
-		$Itemid = \JFactory::getApplication()->input->getInt('Itemid');
+		$Itemid = (int) \JFactory::getApplication()->input->getInt('Itemid', 0);
 		$negId = $Itemid ? -(int) $Itemid : false;
```
Comment (HLeithner) [inline libraries/src/Helper/ModuleHelper.php]:
```suggestion
		$negId = $Itemid ? -$Itemid : false;
```
One case should be enough

File: libraries/src/Application/SiteApplication.php [hunk 3]
```diff
@@ -795,7 +795,7 @@ protected function route()
 		// Execute the parent method
 		parent::route();
 
-		$Itemid = $this->input->getInt('Itemid', null);
+		$Itemid = (int) $this->input->getInt('Itemid', null);
```
Comment (Quy) [inline libraries/src/Application/SiteApplication.php]:
```suggestion
		$Itemid = (int) $this->input->getInt('Itemid', 0);
```

Comment (zero-24) [reply libraries/src/Application/SiteApplication.php]: thanks 👍

**Comment by laoneo** [comment]
I'm using on tons of places getItn since ages, if this is not safe anymore and doesn't always return an integer then we need to inform the extension devs about this change.

**Comment by laoneo** [comment]
Is there a real situation why we need this change? Would be good to have a test case in core to reproduce the issue.
For me it looks strange when we have a function getInt and then have to do an int cast, especially on Itemid which is never an array. It is even more awkward that an array will be casted to integer. Because casting an array to integer doesn't return the value, it returns always 1, so it would be wrong anyway.
Execute `var_dump((int)[2]);`and it will return 1. So if you want to handle that Itemid can be an array then do it right and make a type check and work with arrays properly.

File: administrator/templates/isis/error.php (hunk 1 above)
Comment (SharkyKZ) [reply administrator/templates/isis/error.php]: It's up to you how you do it, but same comment applies to `getInt()`.

**Comment by SharkyKZ** [comment]
>Is there a real situation why we need this change? 

See #26753. This is needed to avoid errors when `Itemid` is being manipulated by user. You're right, though, this should not be cast like that. Should default to 0 or whatever we default to now.

Also test failure is related.

**Comment by SharkyKZ** [comment]
@zero-24 Are changes in FoF files necessary? If so, they should include a note that files have been modified.

**Comment by laoneo** [comment]
But would it then not be better to ensure that Itemid is always an integer before core gets executed instead afterwards?

File: administrator/templates/isis/error.php [hunk 4]
```diff
@@ -27,11 +27,11 @@
...
-$option   = $input->get('option', '');
-$view     = $input->get('view', '');
-$layout   = $input->get('layout', '');
-$task     = $input->get('task', '');
-$itemid   = $input->get('Itemid', 0, 'int');
+$option   = $input->get('option', '', 'cmd');
+$view     = $input->get('view', '', 'cmd');
+$layout   = $input->get('layout', '', 'cmd');
+$task     = $input->get('task', '', 'cmd');
+$itemid   = (int) $input->getInt('Itemid', 0);
```
Comment (bembelimen) [inline administrator/templates/isis/error.php:34]: No need to default ```0```, this is already the case for ```getInt```

**Comment by mbabker** [comment]
> I'm using on tons of places getItn since ages, if this is not safe anymore and doesn't always return an integer then we need to inform the extension devs about this change.

It's still safe to use.  The array support traces back to https://github.com/joomla-framework/filter/pull/13.

Basically since early 2016, stable releases of the `joomla/filter` package have supported cleaning either a single item or an array of items using a given filter.  So `InputFilter::getInstance()->clean(['1', 2, '3', 4], 'int');` should be returning a cleaned array of integers.  Without that you get array to string conversion errors when working with request data that you aren't expecting to be in an array, and there isn't a simple way to deal with that without breaking the inbuilt filtering from the `joomla/input` API on any `get()` call.

> But would it then not be better to ensure that Itemid is always an integer before core gets executed instead afterwards?

I would suggest casting here is the wrong behavior.  The `Itemid` input variable probably needs to be checked early and the application just flat out abort with some kind of Exception if the `Itemid` input variable is an array; an array is flat out invalid data and should be rejected and casting is going to accept it and create potentially unwelcome side effects.  Same goes for other core input variables (i.e. `view`, `format`, `option`, `tmpl`, `layout`, `task` for the "not so legacy legacy" MVC, `type` for the feed views, all of which are the main system variables; at the component level you've got input variables like `id` that should also be checked for array versus single item).  Basically a lack of proper input validation is letting some fields which are expected to be single items come in as arrays which causes problems (BTW the problem isn't the array support in the filter class, it's the fact that the data is coming out of the request superglobals as an array then going into the filter API before there is a chance to even check the data structure is valid).

File: administrator/templates/isis/error.php (hunk 4 above)
Comment (zero-24) [reply administrator/templates/isis/error.php:34]: I know this is just for documentation so we clearly define state that in the code or is there a strong argument against?

**Comment by laoneo** [comment]
Made an alternative https://github.com/joomla/joomla-cms/pull/28751, where the Itemid is correctly set when it is an array.

**Comment by zero-24** [comment]
Awesome thanks @laoneo closing here 👍

**Pull request closed without merging**